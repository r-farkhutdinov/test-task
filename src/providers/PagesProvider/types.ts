import { ReactNode } from 'react';
import { TableData } from '../../types';

export type PagesContextValue = {
    data?: TableData;
    activePage?: string;
    query?: string;
    loading?: boolean;
    error?: string;
    changeActivePage?: (activePage: string) => void;
    setQuery?: (query: string) => void;
    load?: (q?: string) => void;
};

export type Props = {
    children: ReactNode;
}