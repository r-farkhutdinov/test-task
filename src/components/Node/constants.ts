export const basePadding = 22;
export const padding = 16;
export const nodeRootTestId = 'node-container';