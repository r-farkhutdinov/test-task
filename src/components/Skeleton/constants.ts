export const defaultLeftPadding = 16;
export const defaultRightPadding = 32;
export const skeletonRootTestId = 'skeleton';