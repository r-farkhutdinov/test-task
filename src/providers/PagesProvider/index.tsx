import { createContext, useCallback, useEffect, useMemo, useState } from 'react';
import { fetchData } from '../../api';
import type { TableData } from '../../types';
import type { PagesContextValue, Props } from './types';

export const PagesContext = createContext<PagesContextValue>({});

const errorMessage = 'Error fetching data. It is likely that the server is not running. Please start the server from its directory.';

/**
 * Context provider with the core logic of TOC: managing state, fetching data
 */
export const PagesProvider = ({ children }: Props) => {
    const [activePage, setActivePage] = useState<string>();
    const [data, setData] = useState<TableData>();
    const [query, setQuery] = useState('');
    const [error, setError] = useState<string>();
    const [loading, setLoading] = useState(false);

    const load = async (q?: string) => {
        setActivePage(undefined);
        try {
            setLoading(true);
            const result = await fetchData(q);
            setData(result);
        } catch {
            setError(errorMessage);
        } finally {
            setLoading(false);
        }
    };

    // this effect will fire twice in strict mode https://react.dev/reference/react/useEffect 
    useEffect(() => {
        load();
    }, []);

    const changeActivePage = useCallback((page: string) => {
        if (activePage !== page) {
            setActivePage(page);
        }
    }, [activePage]);

    const value = useMemo(() => ({
        activePage,
        data,
        query,
        loading,
        error,
        load,
        setQuery,
        changeActivePage
    }), [activePage, changeActivePage, data, error, loading, query]);

    return (
        <PagesContext.Provider value={value}>{children}</PagesContext.Provider>
    );
};