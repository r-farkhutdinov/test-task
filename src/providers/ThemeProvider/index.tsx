import { createContext, useCallback, useMemo, useState } from 'react';
import { storage } from '../../utils/storage';
import { type ThemeContextValue, type Props, Theme } from './types';

export const themeStorageKey = 'theme';

/**
 * Context provider with theme logic
 */
export const ThemeContext = createContext<ThemeContextValue>({
    theme: Theme.Light,
});

export const ThemeProvider = ({ children }: Props) => {
    const storedTheme = storage.get(themeStorageKey) as Theme;

    const [theme, setTheme] = useState(storedTheme ?? Theme.Light);

    const toggleTheme = useCallback(() => {
        const newTheme = theme === Theme.Light ? Theme.Dark : Theme.Light;
        setTheme(newTheme);
        storage.set(themeStorageKey, newTheme);
    }, [theme]);

    const value = useMemo(() => ({
        theme,
        toggleTheme
    }), [theme, toggleTheme]);

    return (
        <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>
    );
};