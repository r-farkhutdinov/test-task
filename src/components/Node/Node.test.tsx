import { render, screen, fireEvent } from '@testing-library/react';
import { PagesContext } from '../../providers/PagesProvider';
import type { Page } from '../../types';
import { nodeRootTestId } from './constants';
import { Node } from '.';

const mockChangeActivePage = jest.fn();

const defaultProps = {
    id: '1',
    title: 'Node',
    parentId: '2',
    pages: [],
    level: 0,
};

const renderNode = (props: Page = defaultProps) => render(
    <PagesContext.Provider value={{ changeActivePage: mockChangeActivePage }}>
        <Node {...props} />
    </PagesContext.Provider>
);

describe('Node component', () => {
    it('matches snapshot – unexpandable', () => {
        const { asFragment } = renderNode();
        expect(asFragment()).toMatchSnapshot();
    });

    it('matches snapshot – expandable', () => {
        const props = {
            ...defaultProps,
            pages: ['Page']
        };
        const { asFragment } = renderNode(props);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should display title correctly', () => {
        renderNode();
        const nodeDiv = screen.getByTestId(nodeRootTestId);
        expect(nodeDiv).toHaveTextContent('Node');
    });

    it('should expand and collapse on click if expandable', () => {
        const props = {
            ...defaultProps,
            pages: ['Page']
        };
        renderNode(props);
        const nodeDiv = screen.getByTestId(nodeRootTestId);
        fireEvent.click(nodeDiv);
        expect(mockChangeActivePage).toBeCalledWith('1');
    });

    it('should not expand or collapse on click if not expandable', () => {
        renderNode();
        const nodeDiv = screen.getByTestId(nodeRootTestId);
        fireEvent.click(nodeDiv);
        expect(mockChangeActivePage).not.toBeCalled();
    });
});