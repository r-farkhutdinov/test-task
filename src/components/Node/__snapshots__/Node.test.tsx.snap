// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node component should match snapshot – expandable 1`] = `
<DocumentFragment>
  <div
    class="node expandable"
    data-testid="node-container"
    style="padding-left: 22px;"
  >
    <svg
      class="arrow"
      transform="rotate(0)"
    >
      arrow.svg
    </svg>
    Node
  </div>
</DocumentFragment>
`;

exports[`Node component should match snapshot – unexpandable 1`] = `
<DocumentFragment>
  <div
    class="node"
    data-testid="node-container"
    style="padding-left: 42px;"
  >
    Node
  </div>
</DocumentFragment>
`;
